src/org/jitsi/impl/neomedia/transform/csrc/CsrcAudioLevelDispatcher.java

```diff
@@ -82,91 +86,49 @@ public CsrcAudioLevelDispatcher(AudioMediaStreamImpl mediaStream)
         {
-            this.levels = levels;
+            this.levels.set(levels);
 
-            if ((mediaStream != null) && !scheduled)
-            {
-                threadPool.execute(this);
-                scheduled = true;
-            }
-
-            notifyAll();
+            // submit asynchronous delivery of audio levels update
+            threadPool.execute(deliverRunnable);
         }
     }
 
     /**
-     * Waits for new levels to be reported via the <tt>addLevels()</tt> method
-     * and then delivers them to the <tt>AudioMediaStream</tt> that we are
-     * associated with.
+     * Updates associated instance of <tt>AudioMediaStreamImpl</tt> with
+     * current dispatcher
+     * @param mediaStream - new <tt>AudioMediaStreamImpl</tt> to associate
      */
-    @Override
-    public void run()
+    public void setMediaStream(AudioMediaStreamImpl mediaStream)
     {
-        try
-        {
-            do
-            {
-                AudioMediaStreamImpl mediaStream;
-                long[] levels;
-
-                synchronized(this)
-                {
-                    // If the mediaStream is null, this instance is to stop.
-                    mediaStream = this.mediaStream;
-                    if (mediaStream == null)
-                    {
-                        scheduled = false;
-                        break;
-                    }
-
-                    if(this.levels == null)
-                    {
-                        try { wait(); } catch (InterruptedException ie) {}
-                        continue;
-                    }
-                    else
-                    {
-                        levels = this.levels;
-                        this.levels = null;
-                    }
-                }
+        // Reset current media stream and obtain
+        AudioMediaStreamImpl oldStream
+            = this.mediaStream.getAndSet(mediaStream);
 
-                if(levels != null)
-                    mediaStream.audioLevelsReceived(levels);
-            }
-            while (true);
-        }
-        finally
+        if (oldStream != mediaStream)
         {
-            synchronized (this)
-            {
-                scheduled = false;
-            }
+            /*
+             * If the mediaStream changes, it is unlikely that the (audio)
+             * levels are associated with it.
+             */
+            this.levels.set(null);
         }
     }
 
     /**
-     * Causes our run method to exit so that this thread would stop
-     * handling levels.
+     * Delivers last reported audio levels to associated {@link #mediaStream}
      */
-    public void setMediaStream(AudioMediaStreamImpl mediaStream)
+    private void deliverAudioLevelsToMediaStream()
     {
-        synchronized(this)
-        {
-            if (this.mediaStream != mediaStream)
-            {
-                this.mediaStream = mediaStream;
+        final AudioMediaStreamImpl stream = mediaStream.get();
```

Comment:
is there concern that the value of `mediaStream` may be different than one we checked it to schedule the call to `deliverAudioLevelsToMediaStream`?  technically i think this could run in between setting the new media stream in `setMediaStream` and clearing the audio levels later in `setMediaStream`?

Reply:
Good day, @bbaldino!

The scenario you've mentioned became possible after my changes, so that's bad, something need to be changed.

I've had a look at a usage patter of `setMediaStream` and there was 3 usages:
1. Inside constructor of `CsrcAudioLevelDispatcher` to initialize field (safe, because nobody has a reference to `CsrcAudioLevelDispatcher` to cause any kind of races)
2. Inside `CsrcTransformEngine.close()` there was call to `CsrcAudioLevelDispatcher.setMediaStream(null)`, possible race with audio level delivery.
3. Inside `SsrcTransformEngine.close()` there was call to `CsrcAudioLevelDispatcher.setMediaStream(null)`, possible race with audio level delivery.

After having look a usage pattern I've decided to remove method `setMediaStream` and replace it with dedicated `CsrcAudioLevelDispatcher.close()`, and made `MediaStream` final to indicate it does not change.
So after these changes it should be clear that `CsrcAudioLevelDispatcher` has fixed `MediaStream` and dedicated method to close it.

Thank you very much for review, Brian!